### PR TITLE
mozContacts polyfill

### PR DESCRIPTION
**Pull request by mancas**
r? @mcjimenez

### CODE REVIEW

File: polyfills/common/webapi_poly_common.js [hunk 1]
```diff
@@ -137,7 +137,76 @@
...
+    this._fireError = function(error) {
+      if (!_done) {
+        _error = error;
+        _done = true;
+        this.onerror
+          && typeof this.onerror === 'function' && this.onerror(error);
+      }
+    };
+  }
+
```
Comment (mancas) [inline polyfills/common/webapi_poly_common.js:210]: I know we have to update the use of this object in the polyfill of DeviceStorage API. However I'll do that in another PR in order to keep the commits clear.

Comment (AntonioMA) [inline polyfills/common/webapi_poly_common.js:210]: Since I just merged a patch from Carmen that add this to common, go ahead and rebase it :)

File: polyfills/contacts/contacts.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+//  Request: { id: requestId,
+//             data: {
+//             operation: 'oncontactchange'
+//           },
+// Answer: Will be invoked when there's activity with
+// the MozContactChangeEvent (like) object:
+//    { id: requestId,
+//      event: MozContactChangeEvent }
+
+'use strict';
```
Comment (mcjimenez) [inline polyfills/contacts/contacts.js]: Move this to line 35 (inside the function). That way you don't alter what's not in your scope

File: polyfills/contacts/contacts.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    console.log('*-*-*- MozContacts PF: ' + text);
+  }
+
+  // Wishful thinking at the moment...
+  const CONTACTS_SERVICE = 'https://contactsservice.gaiamobile.org';
+
+  // It's nice being monothread...
+  var _currentRequestId = 1;
+
+  function _createAndQueueRequest(data, constructor) {
```
Comment (mcjimenez) [inline polyfills/contacts/contacts.js:46]: Isn't this used like in 3 PF now? (or 4...) ;)

Comment (mcjimenez) [inline polyfills/contacts/contacts.js:46]: Don't move it now, but mark it for a follow up!

File: polyfills/contacts/contacts.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    _oncontactchange: null,
+    _oncontactchangeId: null,
+
+    get oncontactchange() {
+      return this._oncontactchange;
+    },
+
+    set oncontactchange(cb) {
+      this._oncontactchange = cb;
+      this._oncontactchangeId = this._oncontactchangeId || ++_currentRequestId;
```
Comment (mcjimenez) [inline polyfills/contacts/contacts.js:67]: If the handler has already been set, there's no need to set it again. Meaning that this should be something like:
this._oncontactchange = cb;
if (this._oncontactchangeId) {
  return;
}

var self = this;

....
processAnswer: answer => self._oncontactchange && self._oncontactchange(answer.event)

Comment (mcjimenez) [inline polyfills/contacts/contacts.js:67]: That way, once a handler is set we don't need to generate more traffic on the server (which will be mostly a NOP) only to change the local processAnswer

File: polyfills/contacts/contacts.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  function _createAndQueueRequest(data, constructor) {
+    var request = new constructor(++_currentRequestId, data);
+    navConnPromise.
+          then(navConnHelper => navConnHelper.sendObject(request));
+    return request;
+  }
+
+  function _sendRequest(request) {
+    navConnPromise.
+          then(navConnHelper => navConnHelper.sendObject(commandObject));
```
Comment (mcjimenez) [inline polyfills/contacts/contacts.js]: sendObject(request) 
?

File: polyfills/contacts/contacts.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            data: {
+              operation: 'oncontactchange'
+            },
+            processAnswer: answer => cb(answer.event)
+          };
+        }
+      };
+      navConnPromise.
+        then(navConnHelper => navConnHelper.sendObject(commandObject));
+    },
```
Comment (mcjimenez) [inline polyfills/contacts/contacts.js:82]: isn't this _sendRequest(commandObject)?

File: services/contacts/js/app.js [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    var _contacts = navigator.mozContacts;
+    // We can get:
+    // * oncontactchange
+    // * getAll
+    // * methodName
+    var remotePortId = evt.data.remotePortId;
+    var request = evt.data.remoteData;
+    var requestOp = request.data;
+
+    function _cloneContact(mozContact) {
```
Comment (mcjimenez) [inline services/contacts/js/app.js]: Why don't you use the common cloneObject?

File: services/contacts/js/app.js [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+            reason: evt.reason
+          }
+        }
+      });
+    }
+
+    if (requestOp.operation === 'oncontactchange') {
+      _contacts.oncontactchange = listenerTemplate;
+    } else if (requestOp.operation === 'getAll') {
+      var cursor = _contacts.getAll.apply(_contacts, requestOp.params);
```
Comment (mcjimenez) [inline services/contacts/js/app.js]: var cursor = _contacts.getAll(...requestOp.params);
(search for JavaScript spread operator ;))

File: services/contacts/js/app.js [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        _contacts.save(updatedContact).then(() => {
+          sendResult(null);
+        }).catch(error => {
+          sendError(error);
+        });
+      }).catch(error => {
+        sendError(error);
+      });
+    } else {
+      _contacts[requestOp.operation].apply(_contacts, requestOp.params).
```
Comment (mcjimenez) [inline services/contacts/js/app.js]:
```
_contacts[requestOp.operation](...requestOp.params)
```

File: services/contacts/js/app.js [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        then(result => {
+          sendResult(result);
```
Comment (mcjimenez) [inline services/contacts/js/app.js]: then(sendResult).

This should be equivalent

File: services/contacts/js/app.js [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+      }).catch(error => {
+        sendError(error);
```
Comment (mcjimenez) [inline services/contacts/js/app.js]: .catch(sendError)

File: services/contacts/js/app.js [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+      var realContact;
+
+      _contacts.find(filter).then(result => {
+        if (!result.length) {
+          sendError({ name: 'No contacts found' });
+          return;
+        }
+        // Need to update realContact fields
+        var updatedContact = _updateContact(result[0], fakeContact);
+        _contacts.save(updatedContact).then(() => {
```
Comment (mcjimenez) [inline services/contacts/js/app.js]: .then(sendResult) 
or 
.then(() => sendResult(null))
(if null and undefined are not the same for that function)

File: services/contacts/js/app.js (hunk 7 above)
Comment (mancas) [inline services/contacts/js/app.js]: Because we can't clone every object of a XMLHttpRequestEventTarget but we can clone Date objects of a mozContact. That's the main reason. Notice that this function can also clone DOMErrors.

File: polyfills/contacts/contacts.js (hunk 6 above)
Comment (mancas) [inline polyfills/contacts/contacts.js:82]: I've removed that function since it was only used once.

File: services/contacts/js/app.js (hunk 7 above)
Comment (mcjimenez) [inline services/contacts/js/app.js]: I think yet that you should use the common function, (we don't want to have duplicate code) If you don't want it to always make recursive copy just add a new parameter in the common function, do you?

Comment (AntonioMA) [inline services/contacts/js/app.js]: I don't understand your explanation, Manu, sorry. Do you mean we can't clone Date objects? Wouldn't it be better to add this check to the standard clone object so it doesn't bite us on the ass later?

File: services/common/serv_helper.js [hunk 13]
```diff
@@ -41,10 +41,15 @@
...
+  var cloneObject = function(obj, recursive) {
     var cloned = {};
     for (var key in obj) {
       if (typeof obj[key] === 'object') {
+        if (!recursive) {
+          cloned[key] = obj[key];
+          continue;
+        }
+
         cloned[key] = cloneObject(obj[key]);
```
Comment (mcjimenez) [inline services/common/serv_helper.js:53]: It should be better cloned[key] = recursive && cloneObject(obj[key]) || obj[key]
